pkg/providers/version/version.go

```diff
@@ -51,14 +67,77 @@ func (p *Provider) Get(ctx context.Context) (string, error) {
 	if version, ok := p.cache.Get(kubernetesVersionCacheKey); ok {
 		return version.(string), nil
 	}
-	serverVersion, err := p.kubernetesInterface.Discovery().ServerVersion()
-	if err != nil {
-		return "", err
+	var version string
+	var err error
+	// If we're running locally, these environment variables will be empty.
+	host, port := os.Getenv("KUBERNETES_SERVICE_HOST"), os.Getenv("KUBERNETES_SERVICE_PORT")
+	if len(host) == 0 || len(port) == 0 {
+		// If we're running locally, we don't care which APIServer we hit
+		serverVersion, err := p.kubernetesInterface.Discovery().ServerVersion()
+		if err != nil {
+			return "", err
+		}
+		version = fmt.Sprintf("%s.%s", serverVersion.Major, strings.TrimSuffix(serverVersion.Minor, "+"))
+	} else {
+		// If we're running in cluster, we'll ping each APIServer endpoint to get the minimum k8s version.
+		version, err = p.GetMinKubernetesVersion(ctx)
+		if err != nil {
+			return "", err
+		}
 	}
-	version := fmt.Sprintf("%s.%s", serverVersion.Major, strings.TrimSuffix(serverVersion.Minor, "+"))
 	p.cache.SetDefault(kubernetesVersionCacheKey, version)
 	if p.cm.HasChanged("kubernetes-version", version) {
 		logging.FromContext(ctx).With("version", version).Debugf("discovered kubernetes version")
 	}
 	return version, nil
 }
+
+// GetMinKubernetesVersion ensures that we query all known APIServers for the K8s version.
+// This to handle any scenarios where there may be multiple APIServer reporting different
+// K8s versions.
+func (p *Provider) GetMinKubernetesVersion(ctx context.Context) (string, error) {
+	var endpointSlice v1.EndpointSlice
+	if err := p.kubeClient.Get(ctx, types.NamespacedName{Name: "kubernetes", Namespace: "default"}, &endpointSlice); err != nil {
+		return "", fmt.Errorf("getting endpoints, %w", err)
+	}
+	var minVersion *version.Version
+	for _, address := range getAddresses(endpointSlice) {
+		resp, err := p.httpClient.Get(address)
+		if err != nil {
+			return "", fmt.Errorf("sending get request, %w", err)
+		}
+		defer resp.Body.Close()
```

Comment:
Maybe worth a comment similar to the one that we have in the `webhookChecker` just so we don't remove this later on:

```go
		// Close the body to avoid leaking file descriptors
		// Always read the body so we can re-use the connection: https://stackoverflow.com/questions/17948827/reusing-http-connections-in-go

```